HPCC-11151 Support xpath(‘<>’) for write

afishbeck opened the pull request:
When #OPTION('writeInlineContent', true); is used writing a field with
XPATH('Name/<>') will output the string without XML/JSON encoding
the content.  This allows users to insert arbitrary XML/JSON into
the output.

Also works with sets of strings XPATH('SetName/Name/<>').

There are 3 forms:
 '<>' Writes the content without a root tag.  Reading reads all
    content of parent.
 'Name<>' Writes the content without a <Name> tag.  Reading reads the
    content of the tag <Name>.
 'Name/<>' Writes the content inside <Name>.  Reading reads the
    content of the tag <Name>.

Also fixes generated schemas to indicate locations where arbitrary
content might appear.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: fe6fce8eb72d7c71a0b10121d12efb2ff1053a69
Build: failed
Number of Error(s): 6
[  0%] Building CXX object aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/source/client/CoreErrors.cpp.o
[100%] Building CXX object aws-cpp-sdk-sqs/CMakeFiles/aws-cpp-sdk-sqs.dir/source/SQSErrors.cpp.o
/mnt/disk1/home/vamosax/smoketest/PR-10348/HPCC-Platform/common/wuwebview/wuwebview.cpp:912:93: error: no matching function for call to \xe2\x80\x98IResultSetFactory::createResultSetMeta(IConstWUResult*&)\xe2\x80\x99
make[2]: *** [common/wuwebview/CMakeFiles/wuwebview.dir/wuwebview.cpp.o] Error 1
make[1]: *** [common/wuwebview/CMakeFiles/wuwebview.dir/all] Error 2
make: *** [all] Error 2
Get ECLWatch build result

afishbeck committed:
HPCC-11151 Support xpath(‘<>’) for write

When #OPTION('writeInlineContent', true); is used writing a field with
XPATH('Name/<>') will output the string without XML/JSON encoding
the content.  This allows users to insert arbitrary XML/JSON into
the output.

Also works with sets of strings XPATH('SetName/Name/<>').

There are 3 forms:
 '<>' Writes the content without a root tag.  Reading reads all
    content of parent.
 'Name<>' Writes the content without a <Name> tag.  Reading reads the
    content of the tag <Name>.
 'Name/<>' Writes the content inside <Name>.  Reading reads the
    content of the tag <Name>.

Also fixes generated schemas to indicate locations where arbitrary
content might appear.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 66eeb6c909d3c89936b7ff8fc1d1394a15d106da
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 43.914 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |91 |91 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |746 |746 |0 |
|test (thor) |641 |641 |0 |
|test (roxie) |774 |774 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

afishbeck commented:
@richardkchapman you may want to take a look at where this overlapped with your dynamic typeinfo changes.
@ghalliday Can you review?

ghalliday commented on ecl/hql/hqlutil.cpp:
```diff
@@ -8285,8 +8286,21 @@ void extractXmlName(StringBuffer & name, StringBuffer * itemName, StringBuffer *
+            const char *finger = tagName.str() + (lenTagName - lenContents);
+            if (memcmp(finger, XPATH_CONTENTS_TEXT, lenContents) == 0)
+            {
+                bool isNamed = ((lenTagName >= lenContents + 1) && *--finger=='/');
```
minor: finger[-1] cleaner than *--finger, since the latter implies you are going to use the updated variable later on.

ghalliday commented on ecl/hqlcpp/hqlhtcpp.cpp:
```diff
@@ -10923,7 +10930,7 @@ void HqlCppTranslator::buildXmlSerializeSetValues(BuildCtx & ctx, IHqlExpression
         cursor->buildIterateLoop(loopctx, boundCurElement, false);
         OwnedHqlExpr curElement = boundCurElement.getTranslatedExpr();
-        buildXmlSerializeScalar(loopctx, curElement, itemName);
+        buildXmlSerializeScalar(loopctx, curElement, itemName, false);
```
The fourth parameter is a contentFlag, not a boolean.  Using an enum for the parameter would have caught it.

ghalliday commented on common/thorhelper/thorxmlwrite.cpp:
```diff
@@ -117,6 +117,26 @@ void CommonXmlWriter::outputString(unsigned len, const char *field, const char *
     }
 }
 
+void CommonXmlWriter::outputInline(unsigned len, const char *field, const char *fieldname)
```
minor style: Cleaner if length parameter are size32_t rather than unsigned.

ghalliday commented on esp/services/ws_workunits/ws_workunitsService.cpp:
```diff
@@ -3277,7 +3277,7 @@ bool CWsWorkunitsEx::onWUResult(IEspContext &context, IEspWUResultRequest &req,
 
         Owned<DataCacheElement> data;
-        if (!req.getBypassCachedResult())
+        if (0)//!req.getBypassCachedResult())
```
more: Is this left in by mistake?

ghalliday commented on rtl/eclrtl/rtlfield.cpp:
```diff
@@ -679,11 +679,17 @@ size32_t RtlStringTypeInfo::toXML(const byte * self, const byte * selfrow, const
         rtlEStrToStrX(lenAscii, ascii.refstr(), thisLength, str);
-        target.outputString(lenAscii, ascii.getstr(), queryScalarXPath(field));
+        if (field->hasInlineContentXpath())
+            target.outputInline(lenAscii, ascii.getstr(), field->hasNamedContentXpath() ? queryScalarXPath(field) : nullptr);
```
Same issues of conversion to utf8.

ghalliday commented on rtl/eclrtl/rtlfield.cpp:
```diff
@@ -1609,7 +1643,13 @@ size32_t RtlUtf8TypeInfo::toXML(const byte * self, const byte * selfrow, const R
-    target.outputUtf8(thisLength, str, queryScalarXPath(field));
+    if (field->hasInlineContentXpath())
+    {
+        target.outputInline(thisLength, str, field->hasNamedContentXpath() ? queryScalarXPath(field) : nullptr);
```
Need to be careful about length v size.  (Worth naming size variables appropriately.)

ghalliday reviewed (changes requested):
@afishbeck some initial comments.  It is probably worth me reviewing it all again later because it is quite a lot to take in at once.

Main issues are utf8 encoding of strings, and main comment is using enumerations instead of byte.

richardkchapman commented:
@afishbeck This seems to be stalled (and needs rebasing)

afishbeck commented:
Will reopen when I return from PTO.
